### PR TITLE
merge functionality of `io::Sink` into `io::Empty`

### PR DESCRIPTION
**Pull request by vidhanio**
Many times, there is a need for a simple dummy `io::Read`er + `io::Write`r, but currently the only options are `io::Empty` and `io::Sink` respectively. Having both of their functionality together requires writing your own boilerplate for something that makes sense to have in the standard library. This PR adds the functionality of `io::Sink` to `io::Empty`, making `io::Empty` be able to perform the tasks of both of the previous structs. (This idea was first mentioned in #24235)

Note: I also updated some doc comments in `io::utils` in this pull request to fix inconsistencies between `io::Sink` and `io::Empty`.

API Change Proposal: https://github.com/rust-lang/libs-team/issues/49

### CODE REVIEW

File: library/std/src/io/util.rs [hunk 1]
```diff
@@ -19,7 +19,7 @@ pub struct Empty;
 
 /// Constructs a new handle to an empty reader.
 ///
-/// All reads from the returned reader will return <code>[Ok]\(0)</code>.
+/// All reads from the returned reader will return `Ok(0)`.
```
Comment (jyn514) [inline library/std/src/io/util.rs:22]: this removes the intra-doc link, is that intentional?

File: library/std/src/io/util.rs [hunk 2]
```diff
@@ -268,3 +268,154 @@ impl fmt::Debug for Sink {
...
+#[unstable(feature = "io_null", issue = "none")]
+#[non_exhaustive]
+#[derive(Copy, Clone, Default)]
+pub struct Null;
+
+/// Creates an instance of a writer which will successfully consume all data, as
+/// well as a reader which is always at EOF.
+///
+/// All calls to [`write`] on the returned instance will return `Ok(buf.len())`
+/// and the contents of the buffer will not be inspected.
```
Comment (jyn514) [inline library/std/src/io/util.rs:287]: This could be construed to mean that passing uninitialized memory is ok. I think we should remove the second guarantee.
```suggestion
/// All calls to [`write`] on the returned instance will return `Ok(buf.len())`.
```

File: library/std/src/io/util.rs [hunk 3]
```diff
@@ -268,3 +268,154 @@ impl fmt::Debug for Sink {
...
+    #[inline]
+    fn is_write_vectored(&self) -> bool {
+        true
+    }
+
+    #[inline]
+    fn flush(&mut self) -> io::Result<()> {
+        Ok(())
+    }
+}
```
Comment (jyn514) [inline library/std/src/io/util.rs:367]: when would this be useful?

File: library/std/src/io/util.rs [hunk 4]
```diff
@@ -268,3 +268,154 @@ impl fmt::Debug for Sink {
...
+        Ok(0)
+    }
+
+    fn stream_position(&mut self) -> io::Result<u64> {
+        Ok(0)
+    }
+}
+
+#[unstable(feature = "io_null", issue = "none")]
+impl SizeHint for Null {
```
Comment (jyn514) [inline library/std/src/io/util.rs:409]: when would this be useful?

File: library/std/src/io/util.rs [hunk 5]
```diff
@@ -268,3 +268,154 @@ impl fmt::Debug for Sink {
...
+    #[inline]
+    fn upper_bound(&self) -> Option<usize> {
+        Some(0)
+    }
+}
+
+#[unstable(feature = "io_null", issue = "none")]
+impl fmt::Debug for Null {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Null").finish_non_exhaustive()
```
Comment (jyn514) [inline library/std/src/io/util.rs:419]: couldn't this be derived?

File: library/std/src/io/util/tests.rs [hunk 6]
```diff
@@ -124,6 +124,68 @@ fn empty_seeks() {
...
+    assert!(matches!(n.seek(SeekFrom::End(-1)), Ok(0)));
+    assert!(matches!(n.seek(SeekFrom::End(0)), Ok(0)));
+    assert!(matches!(n.seek(SeekFrom::End(1)), Ok(0)));
+    assert!(matches!(n.seek(SeekFrom::End(i64::MAX)), Ok(0)));
+
+    assert!(matches!(n.seek(SeekFrom::Current(i64::MIN)), Ok(0)));
+    assert!(matches!(n.seek(SeekFrom::Current(-1)), Ok(0)));
+    assert!(matches!(n.seek(SeekFrom::Current(0)), Ok(0)));
+    assert!(matches!(n.seek(SeekFrom::Current(1)), Ok(0)));
+    assert!(matches!(n.seek(SeekFrom::Current(i64::MAX)), Ok(0)));
```
Comment (jyn514) [inline library/std/src/io/util/tests.rs:186]: nit: to reduce duplication, maybe use a for loop?

File: library/std/src/lib.rs [hunk 7]
```diff
@@ -340,6 +340,8 @@
 #![feature(thread_local_internals)]
 //
 #![default_lib_allocator]
+//
+#![feature(io_null)]
```
Comment (jyn514) [inline library/std/src/lib.rs:344]: this should go under "only for re-exporting" I think

**Review by jyn514 — COMMENTED** [review]
you added a new submodule `src/rust` that doesn't look intentional.

File: library/std/src/io/util.rs (hunk 3 above)
Comment (vidhanio) [reply library/std/src/io/util.rs:367]: No clue, but it was `impl`'d for `io::Sink`, so I didn't want there to be any weird cases where a user couldn't 1:1 switch from `Sink` to `Null`.

File: library/std/src/io/util.rs (hunk 2 above)
Comment (jyn514) [reply library/std/src/io/util.rs:287]: oh, I see - I probably shouldn't be the one making that decision, but it's a good callout for @joshtriplett.

File: library/std/src/lib.rs (hunk 7 above)
Comment (vidhanio) [reply library/std/src/lib.rs]: Fixed. :)

File: library/std/src/io/util.rs (hunk 5 above)
Comment (vidhanio) [reply library/std/src/io/util.rs]: Derived. :)

File: library/std/src/io/util.rs (hunk 2 above)
Comment (vidhanio) [reply library/std/src/io/util.rs]: I've changed, but will change it back if recommended by someone else. 😄

File: library/std/src/io/util.rs (hunk 1 above)
Comment (vidhanio) [reply library/std/src/io/util.rs]: Oops, I've re-added the the intra-doc links with newer syntax. 👍🏽

File: library/std/src/io/util.rs [hunk 8]
```diff
@@ -19,7 +19,7 @@ pub struct Empty;
 
 /// Constructs a new handle to an empty reader.
 ///
-/// All reads from the returned reader will return `Ok(0)`.
+/// All reads from the returned reader will return [`Ok(0)`].
```
Comment (jyn514) [inline library/std/src/io/util.rs:22]: are you sure this actually works?

Comment (vidhanio) [reply library/std/src/io/util.rs:22]: It seems to work with rust-analyzer's doc parser, but I can add those link aliases at the bottom of the doc if needed.
image

Comment (jyn514) [reply library/std/src/io/util.rs:22]: Hmm, ok - I'd prefer to test that rustdoc works, but I think I vaguely remember a PR allowing this so it's probably fine

Comment (vidhanio) [reply library/std/src/io/util.rs:22]: the rustdoc build failed, and i updated it to use aliasing this morning. :)

File: library/std/src/io/util/tests.rs (hunk 6 above)
Comment (vidhanio) [reply library/std/src/io/util/tests.rs:186]: i just copied the test cases directly from the `io::Sink`/`io::Empty` ones.

**Comment by vidhanio** [comment]
i think that that is a good idea. would it be a good idea to at some point later in time to mark `io::Sink` as deprecated and to move uses of it to `io::Empty`?

**Comment by vidhanio** [comment]
> I think having both Null and Empty might potentially be a bit confusing (especially for people who don't have platform-specific familiarity with /dev/null to make the name seem familiar).

an alternative to this could be to deprecate both `Empty` and `Sink` and move them to a renamed `Null` (if there are any better suggestions).

**Comment by jyn514** [comment]
Please use the existing PR so we have the review context.

**Comment by vidhanio** [comment]
could someone please leave guidance on what to do? `#[unstable]` on just `impl`s is useless, but i need an attribute. should i make it insta-stable?

**Comment by vidhanio** [comment]
i've just noticed this change potentially may be breaking; if users have both `io::Write` and `io::Read` in scope, `io::empty().by_ref` is ambiguous, as `Write::by_ref` and `Read::by_ref` both exist (see log above).

**Comment by jyn514** [comment]
The standard library doesn't consider new trait implementations breaking: https://std-dev-guide.rust-lang.org/code-considerations/breaking-changes/new-trait-impls.html

**Comment by vidhanio** [comment]
@craterbot check

**Comment by jyn514** [comment]
@vidhanio I don't think this needs its own crater check, beta already gets crater runs.

File: library/std/src/io/util.rs [hunk 9]
```diff
@@ -8,24 +8,41 @@ use crate::io::{
     self, BufRead, IoSlice, IoSliceMut, Read, ReadBuf, Seek, SeekFrom, SizeHint, Write,
 };
 
-/// A reader which is always at EOF.
+/// A writer which will move data into the void, and a reader which
+/// is always at EOF.
```
Comment (Mark-Simulacrum) [inline library/std/src/io/util.rs:12]:
```suggestion
/// `Empty` ignores any data written via [`Write`], and will always be empty
/// (returning zero bytes) when read via [`Read`].
```

This wording I think is a little confusing -- maybe we can directly address the trait impls instead?

File: library/std/src/io/util.rs [hunk 10]
```diff
@@ -8,24 +8,41 @@ use crate::io::{
...
+/// This struct is generally created by calling [`empty()`]. Please
+/// see the documentation of [`empty()`] for more details.
 #[stable(feature = "rust1", since = "1.0.0")]
 #[non_exhaustive]
-#[derive(Copy, Clone, Default)]
+#[derive(Copy, Clone, Debug, Default)]
 pub struct Empty;
 
-/// Constructs a new handle to an empty reader.
+/// Creates an instance of a writer which will successfully consume all data, as
```
Comment (Mark-Simulacrum) [inline library/std/src/io/util.rs:21]: I think similarly to my suggestion on Empty, I would rephrase this to more directly talk about a single "thing" that implements two traits and describe that behavior.

Otherwise, this to me sounds like we're returning a writer and a reader (e.g., `fn empty() -> (impl Write, impl Read)` or something), which isn't the case.

File: library/std/src/lib.rs [hunk 11]
```diff
@@ -341,6 +341,7 @@
 #![feature(thread_local_internals)]
 //
 #![default_lib_allocator]
+//
```
Comment (joshtriplett) [inline library/std/src/lib.rs:344]: ?

File: library/std/src/io/util.rs (hunk 2 above)
Comment (joshtriplett) [reply library/std/src/io/util.rs]: @vidhanio Please do keep the comment, to match the types being merged.

**Comment by joshtriplett** [comment]
Posted a couple of comments. r=me with those addressed and the conflict fixed. Thank you!

**Comment by joshtriplett** [comment]
This is, by necessity, insta-stable, so let's confirm we have consensus on it:

@rfcbot merge

File: library/std/src/io/util.rs (hunk 10 above)
Comment (BurntSushi) [reply library/std/src/io/util.rs:21]: Ping @vidhanio. I agree. The wording is a little weird here. Perhaps something like:

> Creates a value that is always at EOF for reads, and ignores all data written.
>
> The value returned implements both the [`Read`] and [`Write`] traits.

**Comment by SUPERCILEX** [comment]
> i think that that is a good idea. would it be a good idea to at some point later in time to mark `io::Sink` as deprecated and to move uses of it to `io::Empty`?

+1 to this, sink should probably be deprecated. Is that something to do in this PR?

**Comment by pitaj** [comment]
@vidhanio please squash your commits so this can be merged

**Comment by vidhanio** [comment]
done @pitaj!

**Comment by SUPERCILEX** [comment]
The since field in the stable impls should probably be updated.
